src/pages/index.js

```diff
@@ -34,50 +36,126 @@ const enableValidation = (formSettings) => {
 
 enableValidation(formSettings);
 
+/* ************************* */
+/* Declaration of Api object */
+/* ************************* */
+const api = new Api({
+  baseUrl: "https://around-api.en.tripleten-services.com/v1",
+  authorization: "a0548a22-3cdf-48eb-9822-7e7daf1b0604",
+});
+
 /* ****************************** */
 /* Declaration of UserInfo object */
 /* ****************************** */
 const userInfo = new UserInfo({
   nameSelector: ".profile__name",
   jobSelector: ".profile__job",
+  avatarSelector: ".profile__picture",
 });
 
 /* ********************** */
 /* Declarations of popups */
 /* ********************** */
+const avatarForm = new PopupWithForm("form[name='avatar-form']", (data) => {
+  avatarForm.submitButton.textContent = "Saving...";
+  api
+    .editProfilePicture(data.link)
+    .then((res) => (res.ok ? true : res.status))
+    .then(() => {
+      userInfo.setAvatar(data.link);
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      avatarForm.submitButton.textContent = "Save";
+      avatarForm.close();
+    });
+});
+avatarForm.setEventListeners();
+
 const cardForm = new PopupWithForm("form[name='card-form']", (data) => {
-  cardGallery.addItem(data);
-  cardForm.close();
+  cardForm.submitButton.textContent = "Saving...";
+  api
+    .addNewCard({ name: data.name, link: data.link })
+    .then((res) => (res.ok ? res.json() : res.status))
+    .then((res) => {
+      cardGallery.addItem(res);
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      cardForm.submitButton.textContent = "Create";
+      cardForm.close();
+    });
 });
 cardForm.setEventListeners();
 
 const profileForm = new PopupWithForm("form[name='profile-form']", (data) => {
-  userInfo.setUserInfo(data);
-  profileForm.close();
+  profileForm.submitButton.textContent = "Saving...";
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```